File: apps/ios/Packages/DropCore/Sources/DropCore/InMemoryNotesRepository.swift
import Foundation

/// 테스트와 SwiftUI 프리뷰용 리포지토리. 네트워크 없이 같은 계약을 지킨다.
public final class InMemoryNotesRepository: NotesRepository, @unchecked Sendable {
    private let lock = NSLock()
    private var notes: [Note]

    /// 실패 경로를 시험하기 위한 손잡이.
    public var loadError: Error?
    public var createError: Error?
    public var mutationError: Error?

    public init(notes: [Note] = []) {
        self.notes = notes
    }

    public func loadNotes() async throws -> [Note] {
        if let loadError { throw loadError }
        return withLock { NoteAssembler.sorted(notes) }
    }

    public func createNote(content: String, parentID: String?) async throws -> Note {
        if let createError { throw createError }
        let note = Note(
            id: UUID().uuidString,
            displayID: withLock { notes.count } + 1,
            content: content,
            parentID: parentID,
            createdAt: Date(),
            updatedAt: Date(),
            source: .mobile
        )
        withLock { notes.insert(note, at: 0) }
        return note
    }

    public func updateNote(id: String, content: String) async throws {
        try mutate(id: id) { $0.replacing(content: content) }
    }

    public func moveToTrash(id: String) async throws {
        try mutate(id: id) { $0.replacing(archivedAt: Optional<Date>.none, deletedAt: Date()) }
    }

    public func restoreFromTrash(id: String) async throws {
        try mutate(id: id) { $0.replacing(deletedAt: Optional<Date>.none) }
    }

    public func archive(id: String) async throws {
        try mutate(id: id) { $0.replacing(archivedAt: Date()) }
    }

    public func unarchive(id: String) async throws {
        try mutate(id: id) { $0.replacing(archivedAt: Optional<Date>.none) }
    }

    public func deletePermanently(id: String) async throws {
        if let mutationError { throw mutationError }
        withLock { notes.removeAll { $0.id == id } }
    }

    public func emptyTrash() async throws {
        if let mutationError { throw mutationError }
        withLock { notes.removeAll { $0.isInTrash } }
    }

    public func setPinned(id: String, isPinned: Bool) async throws {
        try mutate(id: id) { $0.replacing(isPinned: isPinned, pinnedAt: isPinned ? Date() : nil) }
    }

    public func setLocked(id: String, isLocked: Bool) async throws {
        try mutate(id: id) { $0.replacing(isLocked: isLocked) }
    }

    public func setPriority(id: String, priority: Int) async throws {
        try mutate(id: id) { $0.replacing(priority: min(max(priority, 0), 3)) }
    }

    public func updateCategories(id: String, hasLink: Bool, hasMedia: Bool, hasFiles: Bool) async throws {
        try mutate(id: id) { $0.replacing(hasLink: hasLink, hasMedia: hasMedia, hasFiles: hasFiles) }
    }

    private func mutate(id: String, _ transform: (Note) -> Note) throws {
        if let mutationError { throw mutationError }
        withLock {
            guard let index = notes.firstIndex(where: { $0.id == id }) else { return }
            notes[index] = transform(notes[index])
        }
    }

    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}

File: apps/mobile/packages/drop_core/lib/src/in_memory_notes_repository.dart
import 'note.dart';
import 'note_assembler.dart';
import 'notes_repository.dart';

var _idCounter = 0;

String _nextId() {
  _idCounter += 1;
  return '메모리-$_idCounter';
}

/// 테스트와 프리뷰용 리포지토리. 네트워크 없이 같은 계약을 지킨다.
/// DropCore `InMemoryNotesRepository.swift` 대응.
class InMemoryNotesRepository implements NotesRepository {
  final List<Note> _notes;

  /// 실패 경로를 시험하기 위한 손잡이.
  Object? loadError;
  Object? createError;
  Object? mutationError;

  /// `loadNotes`가 실제로 몇 번 불렸는지. 중복 로드를 세기 위한 것.
  int loadCallCount = 0;

  /// 로드를 원하는 시점까지 붙잡아 두기 위한 손잡이.
  /// 겹친 로드를 재현하려면 첫 로드를 여기서 멈춰 세워야 한다.
  Future<void> Function()? beforeLoad;

  /// 생성을 붙잡아 두기 위한 손잡이. 저장이 끝나기 **전** 화면 상태
  /// (낙관적으로 끼워 넣은 노트)를 보려면 여기서 멈춰 세워야 한다.
  Future<void> Function()? beforeCreate;

  InMemoryNotesRepository({List<Note> notes = const []}) : _notes = [...notes];

  @override
  Future<List<Note>> loadNotes() async {
    loadCallCount += 1;
    await beforeLoad?.call();
    final loadError = this.loadError;
    if (loadError != null) throw loadError;
    return NoteAssembler.sorted(_notes);
  }

  @override
  Future<Note> createNote({String content = '', String? parentId}) async {
    await beforeCreate?.call();
    final createError = this.createError;
    if (createError != null) throw createError;
    final note = Note(
      id: _nextId(),
      displayId: _notes.length + 1,
      content: content,
      parentId: parentId,
      createdAt: DateTime.now().toUtc(),
      updatedAt: DateTime.now().toUtc(),
      source: NoteSource.mobile,
    );
    _notes.insert(0, note);
    return note;
  }

  @override
  Future<void> updateNote({required String id, required String content}) =>
      _mutate(id, (note) => note.replacing(content: content));

  @override
  Future<void> moveToTrash(String id) => _mutate(
      id,
      (note) => note.replacing(
          archivedAt: null, deletedAt: DateTime.now().toUtc()));

  @override
  Future<void> restoreFromTrash(String id) =>
      _mutate(id, (note) => note.replacing(archivedAt: null, deletedAt: null));

  @override
  Future<void> archive(String id) => _mutate(
      id, (note) => note.replacing(archivedAt: DateTime.now().toUtc()));

  @override
  Future<void> unarchive(String id) =>
      _mutate(id, (note) => note.replacing(archivedAt: null));

  @override
  Future<void> deletePermanently(String id) async {
    final mutationError = this.mutationError;
    if (mutationError != null) throw mutationError;
    _notes.removeWhere((note) => note.id == id);
  }

  @override
  Future<void> emptyTrash() async {
    final mutationError = this.mutationError;
    if (mutationError != null) throw mutationError;
    _notes.removeWhere((note) => note.isInTrash);
  }

  @override
  Future<void> setPinned(String id, {required bool isPinned}) => _mutate(
      id,
      (note) => note.replacing(
          isPinned: isPinned,
          pinnedAt: isPinned ? DateTime.now().toUtc() : null));

  @override
  Future<void> setLocked(String id, {required bool isLocked}) =>
      _mutate(id, (note) => note.replacing(isLocked: isLocked));

  @override
  Future<void> setPriority(String id, int priority) =>
      _mutate(id, (note) => note.replacing(priority: priority.clamp(0, 3)));

  @override
  Future<void> updateCategories(
    String id, {
    required bool hasLink,
    required bool hasMedia,
    required bool hasFiles,
  }) =>
      _mutate(
          id,
          (note) => note.replacing(
              hasLink: hasLink, hasMedia: hasMedia, hasFiles: hasFiles));

  Future<void> _mutate(String id, Note Function(Note) transform) async {
    final mutationError = this.mutationError;
    if (mutationError != null) throw mutationError;
    final index = _notes.indexWhere((note) => note.id == id);
    if (index < 0) return;
    _notes[index] = transform(_notes[index]);
  }
}
